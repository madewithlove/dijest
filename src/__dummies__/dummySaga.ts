import { put, PutEffect } from 'redux-saga/effects';

export default function*({ type }) {
    yield `type: ${type}`;
    yield put({ type });
}
