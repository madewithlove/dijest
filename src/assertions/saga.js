/**
 * Assert that a saga equals X
 */
export default (actual, effects) => {
    let result;

    // If we provided specific results to test against
    // use those instead of a snapshot
    if (typeof effects !== 'undefined') {
        let next = actual.next();

        return effects.forEach(effect => {
            expect(next.value).toEqual(effect.action);
            next = actual.next(effect.response);
        });
    }

    // Else test against a snapshot
    const actions = [];
    do {
        result = actual.next();
        if (!result.done) {
            actions.push(result);
        }
    } while (!result.done);

    expect(actions).toMatchSnapshot();
};
