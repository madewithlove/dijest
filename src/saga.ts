import { AnyAction } from "redux";
import { call } from "redux-saga/effects";

export type SagaCallback = ((wrapper: SagaWrapper) => void) | any[];

export class SagaWrapper {
    steps = [];

    will(description: string, expected?: any): this {
        return this.then(description, expected);
    }

    then(description: string, expected?: any): this {
        this.steps.push({ description, expected });

        return this;
    }
}

export default function saga(
    description: string,
    saga: any,
    action: AnyAction,
    callback?: SagaCallback,
) {
    const wrapper = new SagaWrapper();

    // Unwrap the saga and store the result of
    // each yield into an array
    const actions = [];
    saga = saga(action);

    let key = 0;
    let result;
    do {
        key++;
        result = saga.next();
        if (!result.done) {
            actions.push(result);
        }
    } while (!result.done);

    // If we passed a callback function, provide the wrapper to it
    if (typeof callback === "function") {
        callback(wrapper);

        return wrapper.steps.forEach(({ description, expected }, key) => {
            it(description, () => {
                if (expected) {
                    expect(actions[key]).toEqual({
                        done: false,
                        value: expected,
                    });
                } else {
                    expect(actions[key]).toMatchSnapshot();
                }
            });
        });
    }

    it(description, () => {
        // If otherwise we passed an array, use it as effects
        if (Array.isArray(callback)) {
            const effects = callback.map(value => ({
                done: false,
                value,
            }));

            return expect(effects).toEqual(actions);
        }

        expect(actions).toMatchSnapshot();
    });
}
