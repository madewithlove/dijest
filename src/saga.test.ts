import { put } from 'redux-saga/effects';
import dummySaga from './__dummies__/dummySaga';
import saga from './saga';

const action = { type: 'INCREMENT' };

saga('can snapshot saga for a given action', dummySaga, action);

saga('can snapshot saga at various points', dummySaga, action, it => {
    it.will('yield some piece of text').then('a put action', put(action));
});

saga('can also snapshot against an array', dummySaga, action, [
    'type: INCREMENT',
    put(action),
]);
