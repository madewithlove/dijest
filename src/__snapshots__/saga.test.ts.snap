// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can snapshot saga for a given action 1`] = `
Array [
  Object {
    "done": false,
    "value": "type: INCREMENT",
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "type": "INCREMENT",
        },
        "channel": null,
      },
    },
  },
]
`;

exports[`yield some piece of text 1`] = `
Object {
  "done": false,
  "value": "type: INCREMENT",
}
`;
